io/fasta/fasta.go

```diff
@@ -66,14 +70,181 @@ type Fasta struct {
+// ParseN does not return EOF if encountered.
+// If an non-EOF error is encountered it returns it and all correctly parsed sequences up to then.
+func (p *Parser) ParseN(maxSequences int) (fastas []Fasta, err error) {
+	for counter := 0; counter < maxSequences; counter++ {
+		fasta, _, err := p.ParseNext()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			return fastas, err
+		}
+		fastas = append(fastas, fasta)
+	}
+	return fastas, nil
+}
+
+// ParseByteLimited parses fastas until byte limit is reached.
+// This is NOT a hard limit. To set a hard limit on bytes read use a
+// io.LimitReader to wrap the reader passed to the Parser.
+func (p *Parser) ParseByteLimited(byteLimit int64) (fastas []Fasta, bytesRead int64, err error) {
+	for bytesRead < byteLimit {
+		fasta, n, err := p.ParseNext()
+		bytesRead += n
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			return fastas, bytesRead, err
+		}
+		fastas = append(fastas, fasta)
+	}
+	return fastas, bytesRead, nil
+}
+
+// ParseNext reads next fasta genome in underlying reader and returns the result
+// and the amount of bytes read during the call.
+// ParseNext only returns an error if it:
+//   - Attempts to read and fails to find a valid fasta sequence.
+//   - Returns reader's EOF if called after reader has been exhausted.
+//   - If a EOF is encountered immediately after a sequence with no newline ending.
+//     In this case the Fasta up to that point is returned with an EOF error.
+//
+// It is worth noting the amount of bytes read are always right up to before
+// the next fasta starts which means this function can effectively be used
+// to index where fastas start in a file or string.
+func (p *Parser) ParseNext() (Fasta, int64, error) {
+	if _, err := p.rd.Peek(1); err != nil {
+		// Early return on error. Probably will be EOF.
+		return Fasta{}, 0, err
+	}
+	// Initialization of parser state variables.
+	var (
+		// Parser looks for a line starting with '>' (U+003E)
+		// that contains the next fasta sequence name.
+		lookingForName = true
+		seqName        string
+		sequence, line []byte
+		err            error
+		totalRead      int64
+	)
+
+	// parse loop begins here.
+	for {
+		line, err = p.rd.ReadSlice('\n')
+		isSkippable := len(line) == 0 || line[0] == ';' // OR short circuits so no panic here.
+		totalRead += int64(len(line))
+		p.line++
+
+		// More general case of error handling.
+		if err != nil {
+			isEOF := errors.Is(err, io.EOF)
+			if isSkippable {
+				if isEOF {
+					// got EOF on a empty or commented line.
+					err = nil
+				}
+				break
+			} else if errors.Is(err, bufio.ErrBufferFull) {
+				// Buffer size too small to read fasta line.
+				return Fasta{}, totalRead, fmt.Errorf("line %d too large for buffer, use larger maxLineSize: %w", p.line+1, err)
```

Comment:
How would we make this fail? I've tried adjusting the parser struct's max line size variable to 1 but still am not able to get this to be thrown.

Reply:
bufio.Reader has a minimum permissible line length of 16. Test now does that, though I feel most of this "error" coverage is in vain. This error in particular can be avoided using bufio.Reader.ReadLine instead of ReadSlice. It just makes parsing harder.

Reply:
That's not to say the maxLineLength argument is in vain, it still provides a way to let users avoid heap allocations during scan by setting an appropiate value.